designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/reconciliation/ScenarioDeploymentReconciler.scala

```diff
@@ -51,4 +87,173 @@ class ScenarioDeploymentReconciler(
+  ): Unit = {
+    if (notFinishedDeploymentsThatAreNotRunning.isEmpty) {
+      logger.info(
+        s"No jobs to recover."
+      )
+    } else {
+      logger.info(
+        s"Starting jobs recovery process. ${notFinishedDeploymentsThatAreNotRunning.size} jobs to recover."
+      )
+    }
+  }
+
+  private def logRecoveryEnd(
+      recoveryResult: List[Try[Unit]]
+  ): Unit = {
+    if (recoveryResult.nonEmpty) {
+      val (successes, failures) = recoveryResult.partition(_.isSuccess)
+      logger.info(
+        s"Jobs recovery process finished. $successes jobs recovered successfully, $failures jobs recovery failed."
+      )
+    }
+  }
+
+  private def collectNotFinishedDeploymentsThatAreNotRunning(
+      processingTypeForWhichJobsShouldBeRecovered: Iterable[ProcessingType]
+  )(implicit user: LoggedUser): Future[List[(ScenarioWithDetailsEntity[CanonicalProcess], DeploymentId)]] = {
+    implicit val freshnessPolicy: DataFreshnessPolicy = Fresh
+    for {
+      // In the perfect World we would fetch all deployments that are not finished for all scenarios, but with deployments modelled as actions, we have to fetch deployed scenarios
+      // (having last state action = deploy and state = finished (not execution finished)). See comment in newdeployment.DeploymentService
+      lastDeployedScenarios <- scenarioRepository.fetchLatestProcessesDetails[CanonicalProcess](
+        ScenarioQuery.deployed.copy(processingTypes = Some(processingTypeForWhichJobsShouldBeRecovered))
+      )
+      notFinishedDeploymentsForScenario = lastDeployedScenarios.map(scenario =>
+        (scenario, DeploymentId.fromActionId(scenario.lastDeployedAction.get.id))
+      )
+      // Currently, job recovery is supported only for DeploymentManagers that supports DeploymentsStatusesQueryForAllScenarios
+      bulkQueriedStatuses <- deploymentStatusesProvider.getBulkQueriedDeploymentStatusesForSupportedManagers(
+        processingTypeForWhichJobsShouldBeRecovered
+      )
+      notFinishedDeploymentsThatAreNotRunning = notFinishedDeploymentsForScenario.filter {
+        case (scenario, deploymentId) =>
+          val bulkQueriedStatusForScenario = bulkQueriedStatuses.getDeploymentStatusesUnsafe(scenario.idData).value
+          !bulkQueriedStatusForScenario.exists(status =>
+            status.deploymentId.contains(deploymentId) && SimpleStateStatus.DefaultFollowingDeployStatuses.contains(
+              status.status
+            )
+          )
+      }
+    } yield notFinishedDeploymentsThatAreNotRunning
+  }
+
+  private def prepareRunDeploymentCommandForValidScenario(
+      scenario: ScenarioWithDetailsEntity[CanonicalProcess],
+      deploymentId: DeploymentId
+  )(implicit user: LoggedUser): Future[Option[(ProcessingType, DMRunDeploymentCommand)]] = {
+    val lastDeployAction = scenario.lastDeployedAction.get
+    // TODO: what should be in user name?
```

Comment:
These TODOs have been copied from the solution that works for embedded DM. After refactors around savepoints and the "redeploy" feature, some of them will be less important or not important at all.